include/raft.h

```diff
@@ -336,6 +309,22 @@ enum {
 
 /**
  * A single RPC message that can be sent or received over the network.
+ *
+ * Updating to a new version of an RPC message is done based on the encoded size
+ * of the message, see e.g. `sizeofRequestVoteV1` in uv_encoding.c
+ *
+ * The RPC message types all have a `reserved` field. When extending such a
+ * struct, and thus shrinking the `reserved` array, one should be careful not to
+ * give meaning the `zero` value because the raft implementation needs to know
+ * whether to other side knows the new field or not. `zero` means the other side
+ * does not know the new field/functionality (this is why raft_tribool exists).
+ * Implementers MUST always zero out unknown fields before passing them to raft.
```

Comment:
I didn't quite realize this when we introduced raft_tribool, but now that I see the `0`-means-older-code-on-the-other-side approach generalized here I'm thinking that perhaps this not the most natural approach.

In most cases `0` is an important and natural value to use. For example let's say that we introduce a new counter field in a message. Do we then have use `0` for "the other side did not include this field", `1` for `0`, `2` for `1`, etc?

Plus, it creates an asymmetry between fields that we already have and the just follow the natural semantics for their values and fields that we add in the future.

Actually we already see the asymmetry in `struct raft_request_vote_result`, where the `vote_granted` field is a regular `bool` and the `pre_vote` field is a `raft_tribool`, which looks weird to me. If we had thought about this pre-vote issue since the very first version of this library, we would certainly have used `bool` not `raft_tribool`, which is something telling. The fact that we use `raft_tribool` now seems a smell that we are mixing two separate concerns: one concern is to store a boolean value in a struct, the other concern is to know whether the other side did include this value in the message it sent us.

This makes me think that a more appropriate approach would be perhaps to separate these two concerns and be explicit about what the other side sent. There are various ways I can imagine to do that, for example adding a `version` field to `struct raft_message` or other variations of that idea. Before we discuss the detail about this possible approach though, I'd like to know how do you feel about this.

Reply:
Yes, I'm certainly open to suggestions. The `zero` approach feels indeed a bit like a 'plumbing' solution.

Reply:
Ok, then my suggestion would be to add an `int version` field to all the various message structures (`struct raft_request_vote`, `struct raft_request_vote_result`,  `struct raft_append_entries`, etc).

Note that I believe this is a bit better than adding the `version` field to `struct raft_message` as I was hinting above, both because what we're versioning is really the specific messages (and not `struct raft_message`), and because as bonus points it should also minimize code changes (since we often pass around the specific messages, not whole `struct raft_message` objects).

This would be pretty much similar to the `version` fields in the `raft_io`/`raft_fsm` structures, making the various message structures effectively "external" structures (borrowing the same terminology that you used in [this comment](https://github.com/canonical/raft/pull/303#issuecomment-1234043078)).

It makes sense considering that those message structures are really part of the `raft_io` interface (specifically the `send()` and `raft_io_recv_cb` parts), so they "inherit" the fact of being external. When receiving messages via `raft_io_recv_cb`, the `raft_io` implementation should be in charge of telling `raft` core what to fields to expect, and it can be done via this new `version` field (like it does for its methods, via the `raft_io->version` field). When sending messages `raft` core should tell `raft_io` what fields it's providing, again via the `version` field.

Regarding encoding and decoding in the libuv-based `raft_io` implementation, we currently use the very first 8 bytes of a message to store the message type (see [here](https://github.com/canonical/raft/blob/09a821f48a827cd75024ad8c689ea7b252ba72b0/src/uv_recv.c#L253) and [here](https://github.com/canonical/raft/blob/09a821f48a827cd75024ad8c689ea7b252ba72b0/src/uv_encoding.c#L213)). I'd suggest to change that slightly, in a fully backward compatible way. We should use the first byte to store the message type (a range of 255 is more than enough even for the future), the second byte to store the version, and the last two bytes are unused and can be reserved for future use. We can decide that version `0` is the current version of all messages, and current servers are already setting the second byte to `0`, so we have backward compatibility.

Note that this also allows us to change the semantics of a message without changing its wire-level length, as at the moment we're using the message length as a "proxy" for its version.

As a side note, with this explicit message versioning approach we should not need to add padding to the various message structures, in the same way we don't use padding for `raft_io`/`raft_fsm`.

To keep things light, in this branch we could just add the `version` field to all the message structures and nothing else. The actual implementation of the encoding and decoding could be done separately. Or we could add the implementation as well, it shouldn't be a lot of code.

Reply:
We could use the first 2 bytes as the message type, the next 2 bytes as version and rest is unused. Then the cast to `unsigned short` [here](https://github.com/canonical/raft/blob/09a821f48a827cd75024ad8c689ea7b252ba72b0/src/uv_encoding.c#L485) should still be fine on old clients with the version set to non-0, or am I missing something?

Reply:
> We could use the first 2 bytes as the message type, the next 2 bytes as version and rest is unused. Then the cast to `unsigned short` [here](https://github.com/canonical/raft/blob/09a821f48a827cd75024ad8c689ea7b252ba72b0/src/uv_encoding.c#L485) should still be fine on old clients with the version set to non-0, or am I missing something?

I'm not sure what you mean. If version is set to non-0, old clients would get a broken message type no? Because they use [`byteFlip64`](https://github.com/canonical/raft/blob/09a821f48a827cd75024ad8c689ea7b252ba72b0/src/uv_recv.c#L253) to get the type.

Unfortunately I believe the only option is to look at the message length, as we already do.

Reply:
We could still modify the current code to use just 2 bytes for the message type, so perhaps down the line we could encode the version in the last 2 bytes, when we are reasonably sure that it's very unlikely to have old clients around.

Reply:
Actually I just realized now what you mean about the cast. I'm not sure the cast helps, since `unsigned short` is at least 16 bits, but it can be longer if the compiler decides to, so no trimming would take place AFAIU.

Reply:
Yeah I assumed it was 2 bytes on all platforms. Practically I think we could do with 4 byte type, 2 byte version and 2 byte reserved/unused, but it would not be portable. I don't think we would ever come across a system with wider than 4-byte unsigned shorts though.

Reply:
> Yeah I assumed it was 2 bytes on all platforms. Practically I think we could do with 4 byte type, 2 byte version and 2 byte reserved/unused, but it would not be portable. I don't think we would ever come across a system with wider than 4-byte unsigned shorts though. I don't think I could get myself to do this, so let's stick with the size approach then.

Agreed.

Reply:
So, practically, we would still version the message structs, but instead of relying on an explicit version in the on wire message, we would rely on the size of the on wire message.

Reply:
> So, practically, we would still version the message structs, but instead of relying on an explicit version in the on wire message, we would rely on the size of the on wire message.

Exactly. As said, I'd also change the code to read only the first 2 bytes to decode the message type, so maybe at some point down the road we'll feel that enough time has passed that it's very unlikely to have both old code and new code servers operating together.

Reply:
> > So, practically, we would still version the message structs, but instead of relying on an explicit version in the on wire message, we would rely on the size of the on wire message.
> 
> Exactly. As said, I'd also change the code to read only the first 2 bytes to decode the message type, so maybe at some point down the road we'll feel that enough time has passed that it's very unlikely to have both old code and new code servers operating together.

I was doing it the other way around, only writing 2 bytes when encoding the message, but that's not strictly needed now, reading only 2 bytes when decoding is what is needed.

Reply:
> > > So, practically, we would still version the message structs, but instead of relying on an explicit version in the on wire message, we would rely on the size of the on wire message.
> > 
> > 
> > Exactly. As said, I'd also change the code to read only the first 2 bytes to decode the message type, so maybe at some point down the road we'll feel that enough time has passed that it's very unlikely to have both old code and new code servers operating together.
> 
> I was doing it the other way around, only writing 2 bytes when encoding the message, but that's not strictly needed now, reading only 2 bytes when decoding is what is needed.

Exactly.

Reply:
>  I'd also change the code to read only the first 2 bytes to decode the message type, so maybe at some point down the road we'll feel that enough time has passed that it's very unlikely to have both old code and new code servers operating together.

I downcast the `uint64_t` type to `uint16_t`, I thought that might be the least intrusive way for now to only use the first 2 bytes of preamble.

Reply:
> > I'd also change the code to read only the first 2 bytes to decode the message type, so maybe at some point down the road we'll feel that enough time has passed that it's very unlikely to have both old code and new code servers operating together.
> 
> I downcast the `uint64_t` type to `uint16_t`, I thought that might be the least intrusive way for now to only use the first 2 bytes of the preamble.

Sounds right to me.